crates/js_api/src/gui/mod.rs

```diff
@@ -0,0 +1,175 @@
+use alloy::primitives::Address;
+use rain_orderbook_app_settings::gui::{
+    Gui, GuiDeployment, GuiFieldDefinition, ParseGuiConfigSourceError,
+};
+use rain_orderbook_bindings::impl_wasm_traits;
+use rain_orderbook_common::dotrain_order::{DotrainOrder, DotrainOrderError};
+use serde::{Deserialize, Serialize};
+use serde_wasm_bindgen::{from_value, to_value};
+use std::collections::HashMap;
+use thiserror::Error;
+use tsify::Tsify;
+use wasm_bindgen::{
+    convert::{
+        js_value_vector_from_abi, js_value_vector_into_abi, FromWasmAbi, IntoWasmAbi,
+        LongRefFromWasmAbi, RefFromWasmAbi, TryFromJsValue, VectorFromWasmAbi, VectorIntoWasmAbi,
+    },
+    describe::{inform, WasmDescribe, WasmDescribeVector, VECTOR},
+    prelude::*,
+};
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Tsify)]
+#[tsify(into_wasm_abi, from_wasm_abi)]
+pub struct TokenDeposit {
+    token: String,
+    amount: String,
+    #[tsify(type = "string")]
+    address: Address,
+}
+impl_wasm_traits!(TokenDeposit);
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Tsify)]
+#[tsify(into_wasm_abi, from_wasm_abi)]
+pub struct FieldValuePair {
+    binding: String,
+    value: String,
+}
+impl_wasm_traits!(FieldValuePair);
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+#[wasm_bindgen]
+pub struct DotrainOrderGui {
+    dotrain_order: DotrainOrder,
+    deployment: GuiDeployment,
+    field_values: HashMap<String, String>,
+    deposits: Vec<TokenDeposit>,
+}
+#[wasm_bindgen]
+impl DotrainOrderGui {
+    #[wasm_bindgen(js_name = "init")]
+    pub async fn init(
+        dotrain: String,
+        deployment_name: String,
+    ) -> Result<DotrainOrderGui, GuiError> {
+        let dotrain_order = DotrainOrder::new(dotrain, None).await?;
+
+        let config = dotrain_order.config();
+        if config.gui.is_none() {
+            return Err(GuiError::GuiConfigNotFound);
+        }
+
+        let gui_config = config.gui.clone().unwrap();
+
+        let gui_deployment = gui_config
+            .deployments
+            .iter()
+            .find(|deployment| deployment.deployment_name == deployment_name)
+            .ok_or(GuiError::DeploymentNotFound(deployment_name))?;
+
+        Ok(Self {
+            dotrain_order,
+            deployment: gui_deployment.clone(),
+            field_values: HashMap::new(),
+            deposits: vec![],
+        })
+    }
+
+    #[wasm_bindgen(js_name = "getGuiConfig")]
+    pub fn get_gui_config(&self) -> Gui {
+        self.dotrain_order.config().gui.clone().unwrap()
+    }
+
+    #[wasm_bindgen(js_name = "getDeposits")]
+    pub fn get_deposits(&self) -> Vec<TokenDeposit> {
+        self.deposits.clone()
+    }
+
+    #[wasm_bindgen(js_name = "saveDeposit")]
+    pub fn save_deposit(&mut self, deposit: TokenDeposit) -> Result<(), GuiError> {
```

Comment:
should this be by ID? Or if we'll force the user of this fn to provide the full TokenDeposit shouldn't we validate that it matches the tokens defined in the gui config?

Reply:
you are right, i'll be implementing some safety checks on this